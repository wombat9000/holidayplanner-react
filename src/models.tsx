export interface AppUser {
    displayName: string;
}