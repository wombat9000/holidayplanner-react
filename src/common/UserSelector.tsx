import * as React from 'react';
import {UserContext} from '../UserContext';
import {MenuItem, Select} from '@material-ui/core';
import {AppUser} from '../models';

export const UserSelector = (props: Props) => {
    const handleChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
        props.setUserHandler({displayName: event.target.value});
    };

    return (
        <UserContext.Consumer>
            {user => {
                const currentUserName = user ? user.displayName : '';
                return (
                    <Select
                        data-qa={'select'}
                        value={currentUserName}
                        onChange={handleChange}
                        className="flex-right">
                        <MenuItem value="">
                            <em>None</em>
                        </MenuItem>
                        <MenuItem value="Wombat">Wombat</MenuItem>
                        <MenuItem value="Moose">Moose</MenuItem>
                    </Select>
                );
            }}
        </UserContext.Consumer>
    );
};

interface Props {
    setUserHandler: (user: AppUser) => void;
}
