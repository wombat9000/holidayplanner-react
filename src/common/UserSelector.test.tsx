import {shallow, ShallowWrapper} from 'enzyme';
import {AppUser} from '../models';
import {UserSelector} from './UserSelector';
import * as React from 'react';

describe('UserSelector', () => {
    let outer: ShallowWrapper<any>;
    const handlerMock: (user: AppUser) => {} = jest.fn();

    beforeEach(() => {
        outer = shallow(<UserSelector setUserHandler={handlerMock}/>);
    });

    it('should match snapshot', () => {
        expect(outer).toMatchSnapshot();
    });

    describe('inside context', () => {
        let InnerContext: (user: AppUser | undefined) => React.ReactElement<void>;

        beforeEach(() => {
            const contextChild: React.ReactNode = outer.props().children;
            InnerContext = contextChild as (user: AppUser | undefined) => React.ReactElement<void>;
        });

        describe('user is present', () => {
            let wrapper: ShallowWrapper<any>;

            beforeEach(() => {
                wrapper = shallow(InnerContext({displayName: 'wombat'}));
            });

            it('should match snapshot', () => {
                expect(wrapper).toMatchSnapshot();
            });

            it('should use username as selected value', () => {
                const select = wrapper.find('[data-qa="select"]');
                const prop = select.prop('value');
                expect(prop).toBe('wombat');
            });

            it('should wire handler into selector', () => {
                const prop = wrapper.find('[data-qa="select"]').prop('onChange');
                const onChangeHandler = prop as (event: any) => void;

                onChangeHandler({target: {value: 'newName'}});

                expect(handlerMock).toHaveBeenCalledWith({displayName: 'newName'});
            });
        });

        describe('user is absent', () => {
            let wrapper: ShallowWrapper<any>;

            beforeEach(() => {
                wrapper = shallow(InnerContext(undefined));
            });

            it('should match snapshot', () => {
                expect(wrapper).toMatchSnapshot();
            });

            it('should use emptystring if user is undefined', () => {
                const select = wrapper.find('[data-qa="select"]');
                const prop = select.prop('value');
                expect(prop).toBe('');
            });
        });
    });
});

/*
// Component.js
const Component = props => (
    <MyContext.Consumer>
        {(context) => (
            <Foo
                bar={props.bar}
                baz={context.baz}
            />
        )}
    </MyContext.Consumer>
);

// Component.test.js
const outer = shallow(<Component bar="bar" />);
const Children = outer.props().children;
const wrapper = shallow(<Children baz="baz" />);

expect(wrapper.find(Foo)).toHaveLength(1);
expect(wrapper.find(Foo)).props().bar).toBe('bar');
expect(wrapper.find(Foo)).props().baz).toBe('baz');
*/