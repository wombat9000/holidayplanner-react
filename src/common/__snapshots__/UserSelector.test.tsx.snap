// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserSelector inside context user is absent should match snapshot 1`] = `
<Select
  IconComponent={[Function]}
  autoWidth={false}
  className="flex-right"
  classes={
    Object {
      "disabled": "MuiSelect-disabled-4",
      "icon": "MuiSelect-icon-5",
      "root": "MuiSelect-root-1",
      "select": "MuiSelect-select-2",
      "selectMenu": "MuiSelect-selectMenu-3",
    }
  }
  data-qa="select"
  displayEmpty={false}
  input={<WithStyles(Input) />}
  multiple={false}
  native={false}
  onChange={[Function]}
  value=""
>
  <WithStyles(MenuItem)
    value=""
  >
    <em>
      None
    </em>
  </WithStyles(MenuItem)>
  <WithStyles(MenuItem)
    value="Wombat"
  >
    Wombat
  </WithStyles(MenuItem)>
  <WithStyles(MenuItem)
    value="Moose"
  >
    Moose
  </WithStyles(MenuItem)>
</Select>
`;

exports[`UserSelector inside context user is present should match snapshot 1`] = `
<Select
  IconComponent={[Function]}
  autoWidth={false}
  className="flex-right"
  classes={
    Object {
      "disabled": "MuiSelect-disabled-4",
      "icon": "MuiSelect-icon-5",
      "root": "MuiSelect-root-1",
      "select": "MuiSelect-select-2",
      "selectMenu": "MuiSelect-selectMenu-3",
    }
  }
  data-qa="select"
  displayEmpty={false}
  input={<WithStyles(Input) />}
  multiple={false}
  native={false}
  onChange={[Function]}
  value="wombat"
>
  <WithStyles(MenuItem)
    value=""
  >
    <em>
      None
    </em>
  </WithStyles(MenuItem)>
  <WithStyles(MenuItem)
    value="Wombat"
  >
    Wombat
  </WithStyles(MenuItem)>
  <WithStyles(MenuItem)
    value="Moose"
  >
    Moose
  </WithStyles(MenuItem)>
</Select>
`;

exports[`UserSelector should match snapshot 1`] = `<[object Object] />`;
